### PR TITLE
Consolidate all cabal.project files

### PR DESCRIPTION
**Pull request by pepeiborra**
Consolidate all the various project files into a single one that aggregates all the `allow-newer` and `constraints` from the existing project files.
- `cabal.project`
- `cabal-ghc90.project` 
- `cabal-ghc92.project` 

This is possible now that our dependencies (stylish-haskell, hlint, etc.) have moved forward to `ghc-lib` 9.2 and that all the per-ghc-version cabal flags have been eliminated (#2867).

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2866"></a>

### CODE REVIEW

File: cabal.project
Comment (pepeiborra) [reply cabal.project:52]: WIP - I am looking to delete `cabal-ghc92.project` as well

File: .github/workflows/test.yml
Comment (michaelpj) [inline .github/workflows/test.yml:117]: Previously we could have had plugins which _built_ on GHC X (and hence weren't commented out in the special `cabal.project`), but whose _tests_ didn't pass (and hence weren't included in the testing workflow. These plugins would have been built by this step, which would have ensured that they kept compiling, even if the tests didn't pass.

Now I think we will only build stuff that's necessary for the tests that we try and run, so if we don't test a plugin we also won't ensure that it keeps building.

Also I think we won't check compilation of the main HLS executable either?

Comment (pepeiborra) [reply .github/workflows/test.yml]: This is a good observation, thanks! 

I deleted this step before figuring out that we can use the `buildable` field to selectively disable projects per ghc version. With that trick, hopefully this step can stay and we can avoid all the pitfalls that you discovered

Comment (michaelpj) [reply .github/workflows/test.yml]: I have a feeling that that won't work. I think I suggested this to @jneira in the past and he convinced me that it wouldn't work. I can't remember why though >:(

Comment (pepeiborra) [reply .github/workflows/test.yml]: I can't see the reason why it wouldn't work. CI is failing because the Cabal plan it computes is wrong for 9.0.1, but I get the right plan locally so it might be a CI bug

Comment (michaelpj) [reply .github/workflows/test.yml]: I think maybe it had something to do with Cabal not taking `buildable` into account when making a plan, so it won't be clever enough to try setting the `splice` flag to `false` if the splice plugin isn't buildable, it'll just fail. Worth checking though.

Comment (pepeiborra) [reply .github/workflows/test.yml]: It doesn't need to be:

https://github.com/haskell/haskell-language-server/blob/4b475b0adf665d159b1de54f260eeb84bfdf58fe/haskell-language-server.cabal#L237-L239

Comment (michaelpj) [reply .github/workflows/test.yml]: Ah, so you're going to keep the existing machinery? Maybe the thing that wouldn't work was a more ambitious thing: I'd though that maybe we could *automatically* have plugins be enabled or disabled depending on whether the plugin package was buildable given the rest of the configuration or not. So then we'd only have to put the logic for deciding buildability in the plugin package, which would be nice.

Comment (pepeiborra) [reply .github/workflows/test.yml]: One step at a time..

**Comment by pepeiborra** [comment]
@michaelpj the Nix CI is failing, but I am going to leave it to you (after this PR has landed). Deal?

File: plugins/hls-brittany-plugin/hls-brittany-plugin.cabal
Comment (fendor) [inline plugins/hls-brittany-plugin/hls-brittany-plugin.cabal:23]: This is not ideal, afaict, this spreads disabling plugins over multiple files, but it looks like we have a hard time doing better than this.
Is https://github.com/haskell/cabal/pull/7783 good enough to allow us to centralise this again?

**Review by fendor — APPROVED** [review]
LGTM, awesome!

File: plugins/hls-brittany-plugin/hls-brittany-plugin.cabal
Comment (pepeiborra) [reply plugins/hls-brittany-plugin/hls-brittany-plugin.cabal:23]: I don't know whether it's good enough, but I agree that buildabilty with ghc x.y would sit better in the project file

**Comment by michaelpj** [comment]
> the Nix CI is failing

It's already broken, you're not making it any worse!

File: plugins/hls-brittany-plugin/hls-brittany-plugin.cabal
Comment (michaelpj) [reply plugins/hls-brittany-plugin/hls-brittany-plugin.cabal:23]: I actually think it's better in the plugin files! "Does this plugin work on GHC X" feels like a concern of the plugin.

File: cabal-ghc92.project
Comment (michaelpj) [inline cabal-ghc92.project:52]: Shouldn't the retrie, splice, and tactics plugins also get `buildable` conditionals?

Comment (michaelpj) [reply cabal-ghc92.project:52]: In fact, how is the CI passing?? Did they just start working without us noticing?

Comment (michaelpj) [reply cabal-ghc92.project:52]: Oh, it's because we still have the GHC conditionals for inclusion in `haskell-language-server.cabal`

Comment (pepeiborra) [reply cabal-ghc92.project:52]: Yeah you are right - I'm not sure how `cabal build` decides what to build, so far I'm just adding the `buildable` conditionals on demand

**Pull request merged**
